beacon-chain/sync/subscriber.go

```diff
@@ -0,0 +1,91 @@
+package sync
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+)
+
+// subHandler represents handler for a given subscription.
+type subHandler func(context.Context, proto.Message) error
+
+func notImplementedSubHandler(_ context.Context, _ proto.Message) error {
+	return errors.New("not implemented")
+}
+
+// validator should verify the contents of the message, propagate the message
+// as expected, and return true or false to continue the message processing
+// pipeline.
+type validator func(context.Context, proto.Message, p2p.Broadcaster) bool
+
+// noopValidator is a no-op that always returns true and does not propagate any
+// message.
+func noopValidator(_ context.Context, _ proto.Message, _ p2p.Broadcaster) bool {
+	return true
+}
+
+// subscribe to a given topic with a given validator and subscription handler.
+// The base protobuf message is used to initialize new messages for decoding.
+func (r *RegularSync) subscribe(topic string, v validator, h subHandler) {
+	base := p2p.GossipTopicMappings[topic]
+	if base == nil {
+		panic(fmt.Sprintf("%s is not mapped to any message in GossipTopicMappings", topic))
+	}
+
+	topic += r.p2p.Encoding().ProtocolSuffix()
+	log := log.WithField("topic", topic)
+
+	sub, err := r.p2p.PubSub().Subscribe(topic)
+	if err != nil {
+		panic(err)
```

Comment:
should it panic here ?

Reply:
Yeah, this can only panic on mis-configuration.

Reply:
Added comment

Reply:
thanks